### ISSUE TITLE
Command Line -d1 not working anymore in 1.28.0.0

### ISSUE THREAD

**Issue by 8bit-Dude** [AppleWin/AppleWin#663]
This seems to be an issue that has re-appeared. Using -d1 with 1.28.0.0 does not work, a message appears saying the the floppy failed to insert.

**Comment by tomcw** [comment]
I can't reproduce this. I tested 1.28.0.0 and the latest 1.29.0.0.

Can you attach the AppleWin.log file?
EG: (but use the _exact command line_ that fails for you)
```
> applewin.exe -log -d1 "myfile.dsk"
```

NB. You only need quotes if the pathname has spaces (but it doesn't hurt to always use quotes).

NB. For every release (since April 2018 / AppleWin 1.27.1.0), I do test the -d1 (and related switches).

**Comment by Archange427** [comment]
I have already encountered the same issue. I think it's a directory problem:
- if I use a launch.bat with this line: C:\Progx86\AppleWin\applewin.exe -d1 "test.dsk",
Even if test.dsk IS IN THE SAME DIRECTORY as launch.bat, AppleWin is unable to open it...

- if I set the full directory for test.dsk in launch.bat
(->C:\Progx86\AppleWin\applewin.exe -d1 "D:\TEMP\test.dsk"), 
it's OK,

Dunno if it's the same kind of configuration used by @8bit-Dude ...

Issue with 1.28.x version.
Seems that AppleWin doesn't use the current directory properly...
I even did a test with "test.dsk" in the AppleWin directory and 
C:\Progx86\AppleWin\applewin.exe -d1 "test.dsk" doesnt' open it either.

**Comment by 8bit-Dude** [comment]
I can confirm the problem is same for me: relative directories (../../disks/game.dsk) are not working any longer since release 1.28.x

**Comment by tomcw** [comment]
And what version of 1.27.x were you using when it was working?
I'm trying to pin down when this regression occurred.

**Comment by Archange427** [comment]
Hi Tom.
I did some quick tests. If I use my batch file again:
- 1.27.0.0 => WORKING
- 1.27.1.0 => NO MORE WORKING
and the same (not working) for later 1.27.x (27.2, 27.6, 27.13) versions that I tested...

**Comment by tomcw** [comment]
Thanks Arnaud - I'll take a look at the difference between those 2 versions.

**Comment by tomcw** [comment]
Triage:
- process cmd line and keep copies of the -d1|d2|h1|h2 pathnames (which may be relative or absolute)
- main loop calls `LoadConfiguration()`
  - **// Harddisk**
  - from Registry load REGVALUE_PREF_HDV_START_DIR "HDV Starting Directory"
  - call `SetCurrentImageDir()` with this Registry value
    - copy to `g_sCurrentDir`
    - `SetCurrentDirectory(g_sCurrentDir)`
  - `HD_LoadLastDiskImage(HARDDISK_1)`
  - `HD_LoadLastDiskImage(HARDDISK_2)`
  - **// Floppy**
  - from Registry load REGVALUE_PREF_START_DIR "Starting Directory"
  - call `SetCurrentImageDir()` with this Registry value
    - copy to `g_sCurrentDir`  **-- NB. stomps on g_sCurrentDir set by HDV above**
    - `SetCurrentDirectory(g_sCurrentDir)`
  - `sg_Disk2Card.LoadLastDiskImage(DRIVE_1)`
  - `sg_Disk2Card.LoadLastDiskImage(DRIVE_2)`
- `InsertFloppyDisks()`
- `InsertHardDisks()`

**// InsertFloppyDisks()**
- DoDiskInsert()
  - GetFullPath(szFileName)
    - if abs, then return szFileName
    - if rel, then return GetCurrentDirectory() + szFileName

Basically passing `-d1 <relative pathname>` will result in AppleWin trying to load the floppy image from Registry REGVALUE_PREF_START_DIR "Starting Directory" + `<relative pathname>`; when what the user expects is to be relative to the current cmd prompt.

EG: all these should work:

```
C:\AppleII> applewin\applewin -d1 images\image.dsk
C:\AppleII\AppleWin> applewin -d1 ..\images\image.dsk
C:\AppleII\Images> ..\applewin\applewin -d1 image.dsk
```

**Comment by tomcw** [comment]
Here's my proposed fix for this issue:

1. For all the image loading switches (-d1,-d2,-h1,-h2,-s5d1,-s5d2,etc), then if passed a relative pathname, then treat relative to the current directory, eg. (from above), these **will now** work:

```
C:\AppleII> applewin\applewin -d1 images\image.dsk
C:\AppleII\AppleWin> applewin -d1 ..\images\image.dsk
C:\AppleII\Images> ..\applewin\applewin -d1 image.dsk
```

2. All image loading switches **will now** internally set the `g_sCurrentDir` and also call `SetCurrentDirectory()`.

NB. From one version of AppleWin to the next, it's not guaranteed what order these switches are processed, but currently the lower slots are done first; and for each slot, the lower drive is done first. But I wouldn't say this is "spec", so don't rely on it!

3. To compensate for (2), there's a new switch `-current-dir <path>` which is guaranteed to be processed after all the image loading switches. This too sets the `g_sCurrentDir` and also call `SetCurrentDirectory()`.

Use this switch if you have a mix of (eg) -d1 and -h1 loading images from different folders, and you want to guarantee the current dir.

So (1) & (2) fix #663 and #691.

---

Another thing to bear in mind:
- After successfully saving or loading a save-state, this updates the Registry key `Save State Filename` (`REGVALUE_SAVESTATE_FILENAME`) with the newly saved/loaded pathname.
- Then on F12 (load save-state), it will use this to set `g_strSaveStatePath`, and open the load dialog showing this folder.

I think this is fine, but it may look odd if you hit F11 (and cancel) then hit F12... as the folders may be different.

**Comment by tomcw** [comment]
Added new tests here: https://github.com/AppleWin/AppleWin-Test/commit/86575b18c3ee351b3e1dfa11dd64dd8c15603f29

**Comment by tomcw** [comment]
@8bit-Dude , @Archange427 - there's a new AppleWin 1.29.14.0 [here](https://github.com/AppleWin/AppleWin/releases/download/v1.29.14.0/AppleWin1.29.14.0.zip), which contains an update for this issue. Can you give it a try?

**Comment by Archange427** [comment]
it seems OK with my own use (I only tested "-d1" for now)
Thank you for the update.

**Comment by tomcw** [comment]
Closing this old bug.
